In [1]:
import pandas as pd
import numpy as np
%matplotlib inline
import matplotlib.pyplot as plt
import seaborn as sns

import plotly.express as px





#plt.figure(figsize = (17,8))

In [2]:
customers = pd.read_csv('olist_customers_dataset.csv')
orders = pd.read_csv('olist_orders_dataset.csv')
order_items = pd.read_csv('olist_order_items_dataset.csv')

In [3]:
customers.head(5) #таблица с уникальными идентификаторами пользователей

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [4]:
#customer_id — позаказный идентификатор пользователя

#customer_unique_id —  уникальный идентификатор пользователя  (аналог номера паспорта)

#customer_zip_code_prefix —  почтовый индекс пользователя

#customer_city —  город доставки пользователя

#customer_state —  штат доставки пользователя

In [5]:
orders.head(5) #таблица заказов

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [6]:
#order_id —  уникальный идентификатор заказа (номер чека)

#customer_id —  позаказный идентификатор пользователя

#order_status —  статус заказа

#order_purchase_timestamp —  время создания заказа

#order_approved_at —  время подтверждения оплаты заказа

#order_delivered_carrier_date —  время передачи заказа в логистическую службу

#order_delivered_customer_date —  время доставки заказа

#order_estimated_delivery_date —  обещанная дата доставки

In [7]:
order_items.head(5) #товарные позиции, входящие в заказы

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [8]:
#order_id —  уникальный идентификатор заказа (номер чека)

#order_item_id —  идентификатор товара внутри одного заказа

#product_id —  ид товара (аналог штрихкода)

#seller_id — ид производителя товара

#shipping_limit_date —  максимальная дата доставки продавцом для передачи заказа партнеру по логистике

#price —  цена за единицу товара

#freight_value —  вес товара

# 1.СКОЛЬКО У НАС ПОЛЬЗОВАТЕЛЕЙ, КОТОРЫЕ СОВЕРЩИЛИ ПОКУПКУ ТОЛЬКО ОДИН РАЗ?

In [10]:
df=pd.merge(customers, orders, on='customer_id', how='inner')

#соеденили данные о заказах с айди пользователей

In [11]:
df

customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

       customer_zip_code_prefix          customer_city customer_state  \
0                         14409                 franca             SP   
1                          9790  sao bernardo do campo             SP   
2                          1151              sao paulo             SP   
3                          8775        mogi das cruzes             SP   
4                         13056               campinas             SP   
...                         ...                    ...            ...   
99436                      3937              sao paulo             SP   
99437                      6764        taboao da serra             SP   
99438                     60115              fortaleza             CE   
99439                     92120                 canoas             RS   
99440                      6703                  cotia             SP   

                               order_id order_status order_purchase_timestamp  \
0      00e7ee1b050b8499577073aeb2a297a1    delivered      2017-05-16 15:05:35   
1      29150127e6685892b6eab3eec79f59c7    delivered      2018-01-12 20:48:24   
2      b2059ed67ce144a36e2aa97d2c9e9ad2    delivered      2018-05-19 16:07:45   
3      951670f92359f4fe4a63112aa7306eba    delivered      2018-03-13 16:06:38   
4      6b7d50bd145f6fc7f33cebabd7e49d0f    delivered      2018-07-29 09:51:30   
...                                 ...          ...                      ...   
99436  6760e20addcf0121e9d58f2f1ff14298    delivered      2018-04-07 15:48:17   
99437  9ec0c8947d973db4f4e8dcf1fbfa8f1b    delivered      2018-04-04 08:20:22   
99438  fed4434add09a6f332ea398efd656a5c    delivered      2018-04-08 20:11:50   
99439  e31ec91cea1ecf97797787471f98a8c2    delivered      2017-11-03 21:08:33   
99440  28db69209a75e59f20ccbb5c36a20b90    delivered      2017-12-19 14:27:23   

         order_approved_at order_delivered_carrier_date  \
0      2017-05-16 15:22:12          2017-05-23 10:47:57   
1      2018-01-12 20:58:32          2018-01-15 17:14:59   
2      2018-05-20 16:19:10          2018-06-11 14:31:00   
3      2018-03-13 17:29:19          2018-03-27 23:22:42   
4      2018-07-29 10:10:09          2018-07-30 15:16:00   
...                    ...                          ...   
99436  2018-04-07 16:08:45          2018-04-11 02:08:36   
99437  2018-04-04 08:35:12          2018-04-05 18:42:35   
99438  2018-04-08 20:30:03          2018-04-09 17:52:17   
99439  2017-11-03 21:31:20          2017-11-06 18:24:41   
99440  2017-12-19 18:50:39          2017-12-21 19:17:21   

      order_delivered_customer_date order_estimated_delivery_date  
0               2017-05-25 10:35:35           2017-06-05 00:00:00  
1               2018-01-29 12:41:19           2018-02-06 00:00:00  
2               2018-06-14 17:58:51           2018-06-13 00:00:00  
3               2018-03-28 16:04:25           2018-04-10 00:00:00  
4               2018-08-09 20:55:48           2018-08-15 00:00:00  
...                             ...                           ...  
99436           2018-04-13 20:06:37           2018-04-25 00:00:00  


In [12]:
order_counts = df.groupby('customer_unique_id')['order_id'].count().reset_index()


In [13]:
order_counts

customer_unique_id  order_id
0      0000366f3b9a7992bf8c76cfdf3221e2         1
1      0000b849f77a49e4a4ce2b2a4ca5be3f         1
2      0000f46a3911fa3c0805444483337064         1
3      0000f6ccb0745a6a4b88665a16c9f078         1
4      0004aac84e0df4da2b147fca70cf8255         1
...                                 ...       ...
96091  fffcf5a5ff07b0908bd4e2dbc735a684         1
96092  fffea47cd6d3cc0a88bd621562a9d061         1
96093  ffff371b4d645b6ecea244b27531430a         1
96094  ffff5962728ec6157033ef9805bacc48         1
96095  ffffd2657e2aad2907e67c3e9daecbeb         1

[96096 rows x 2 columns]

In [14]:
one_time_customers = order_counts[order_counts['order_id'] == 1]

In [15]:
one_time_customers

customer_unique_id  order_id
0      0000366f3b9a7992bf8c76cfdf3221e2         1
1      0000b849f77a49e4a4ce2b2a4ca5be3f         1
2      0000f46a3911fa3c0805444483337064         1
3      0000f6ccb0745a6a4b88665a16c9f078         1
4      0004aac84e0df4da2b147fca70cf8255         1
...                                 ...       ...
96091  fffcf5a5ff07b0908bd4e2dbc735a684         1
96092  fffea47cd6d3cc0a88bd621562a9d061         1
96093  ffff371b4d645b6ecea244b27531430a         1
96094  ffff5962728ec6157033ef9805bacc48         1
96095  ffffd2657e2aad2907e67c3e9daecbeb         1

[93099 rows x 2 columns]

In [16]:
count_one_time_customers = len(one_time_customers)

In [17]:
count_one_time_customers 

93099

# ОТВЕТ 93099 

# 2 Сколько заказов в месяц в среднем не доставляется по разным причинам (вывести детализацию по причинам)?

In [19]:
#заказ будет считаться недоставленным в случае отмены canceled , недоступен unavailable , 
#а также просрочки доставки order_estimated_delivery_date

In [20]:
not_delivered_orders = orders[orders['order_status'] != 'delivered'] 
#Все недоставленные заказы (в том числе в пути без нарушения сроков)

In [21]:
not_delivered_orders

order_id                       customer_id  \
6      136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
44     ee64d42b8cf066f35eac1cf57de1aa85  caded193e8e47b8362864762a83db3c5   
103    0760a852e4e9d89eb77bf631eaaf1c84  d2a79636084590b7465af8ab374a8cf5   
128    15bed8e2fec7fdbadb186b57c46c92f2  f3f0e613e0bdb9c7cee75504f0f90679   
154    6942b8da583c2f9957e990d028607019  52006a9383bf149a4fb24226b173106f   
...                                 ...                               ...   
99283  3a3cddda5a7c27851bd96c3313412840  0b0d6095c5555fe083844281f6b093bb   
99313  e9e64a17afa9653aacf2616d94c005b8  b4cd0522e632e481f8eaf766a2646e86   
99347  a89abace0dcc01eeb267a9660b5ac126  2f0524a7b1b3845a1a57fcf3910c4333   
99348  a69ba794cc7deb415c3e15a0a3877e69  726f0894b5becdf952ea537d5266e543   
99415  5fabc81b6322c8443648e1b21a6fef21  32c9df889d41b0ee8309a5efb6855dcb   

      order_status order_purchase_timestamp    order_approved_at  \
6         invoiced      2017-04-11 12:22:08  2017-04-13 13:25:17   
44         shipped      2018-06-04 16:44:48  2018-06-05 04:31:18   
103       invoiced      2018-08-03 17:44:42  2018-08-07 06:15:14   
128     processing      2017-09-03 14:22:03  2017-09-03 14:30:09   
154        shipped      2018-01-10 11:33:07  2018-01-11 02:32:30   
...            ...                      ...                  ...   
99283     canceled      2018-08-31 16:13:44                  NaN   
99313   processing      2018-01-05 23:07:24  2018-01-09 07:18:05   
99347     canceled      2018-09-06 18:45:47                  NaN   
99348  unavailable      2017-08-23 16:28:04  2017-08-28 15:44:47   
99415  unavailable      2017-10-10 10:50:03  2017-10-14 18:35:57   

      order_delivered_carrier_date order_delivered_customer_date  \
6                              NaN                           NaN   
44             2018-06-05 14:32:00                           NaN   
103                            NaN                           NaN   
128                            NaN                           NaN   
154            2018-01-11 19:39:23                           NaN   
...                            ...                           ...   
99283                          NaN                           NaN   
99313                          NaN                           NaN   
99347                          NaN                           NaN   
99348                          NaN                           NaN   
99415                          NaN                           NaN   

      order_estimated_delivery_date  
6               2017-05-09 00:00:00  
44              2018-06-28 00:00:00  
103             2018-08-21 00:00:00  
128             2017-10-03 00:00:00  
154             2018-02-07 00:00:00  
...                             ...  
99283           2018-10-01 00:00:00  
99313           2018-02-06 00:00:00  
99347           2018-09-27 00:00:00  
99348           2017-09-15 00:00:00  
99415           2017-10-23 00:00:00  

[2963 rows x 8 columns]

In [22]:
#отмененные заказы 

In [23]:
canceled_orders = orders[orders['order_status'] == 'canceled']

In [24]:
canceled_orders_ans = len(canceled_orders)

In [25]:
canceled_orders_ans

625

In [26]:
print(f'Количество недоставленных заказов из-за статуса "отменен": {canceled_orders_ans}')

Количество недоставленных заказов из-за статуса "отменен": 625


In [27]:
#недоступные заказы 

In [28]:
unavailable_orders = orders[orders['order_status'] == 'unavailable']

In [29]:
unavailable_orders_ans = len(unavailable_orders)

In [30]:
print(f'Количество недоставленных заказов из-за статуса "недоступен": {unavailable_orders_ans}')

Количество недоставленных заказов из-за статуса "недоступен": 609


In [31]:
#также заказы могут быть не доставлены из-за нарушения сроков
#order_purchase_timestamp —  время создания заказа

#order_approved_at —  время подтверждения оплаты заказа

#order_delivered_carrier_date —  время передачи заказа в логистическую службу

#order_delivered_customer_date —  время доставки заказа

#order_estimated_delivery_date —  обещанная дата доставки

In [32]:
orders = orders.astype({'order_estimated_delivery_date':'datetime64',
                                'order_delivered_customer_date':'datetime64',
                                'order_delivered_carrier_date':'datetime64',
                                'order_approved_at': 'datetime64',
                                'order_purchase_timestamp': 'datetime64'})

In [57]:
orders.dtypes

order_id                                  object
customer_id                               object
order_status                              object
order_purchase_timestamp          datetime64[ns]
order_approved_at                 datetime64[ns]
order_delivered_carrier_date      datetime64[ns]
order_delivered_customer_date     datetime64[ns]
order_estimated_delivery_date     datetime64[ns]
delivered_delta                  timedelta64[ns]
dtype: object

In [34]:
primer_orders = orders[orders['order_id'] == '00143d0f86d6fbd9f9b38ab440ac16f5']


In [35]:
primer_orders

order_id                       customer_id  \
36265  00143d0f86d6fbd9f9b38ab440ac16f5  2e45292979b9b2700ea09560eeb0f803   

      order_status order_purchase_timestamp   order_approved_at  \
36265    delivered      2017-10-16 15:29:43 2017-10-16 16:07:52   

      order_delivered_carrier_date order_delivered_customer_date  \
36265          2017-10-16 23:04:43           2017-10-27 18:43:56   

      order_estimated_delivery_date  
36265                    2017-11-06

In [36]:
orders['delivered_delta'] = orders.order_delivered_customer_date - orders.order_estimated_delivery_date 
#есди значение положительное, значит случилась просрочка доставки)


In [37]:
orders

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp   order_approved_at  \
0        delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   
...            ...                      ...                 ...   
99436    delivered      2017-03-09 09:54:05 2017-03-09 09:54:05   
99437    delivered      2018-02-06 12:58:58 2018-02-06 13:10:37   
99438    delivered      2017-08-27 14:46:43 2017-08-27 15:04:16   
99439    delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
99440    delivered      2018-03-08 20:57:30 2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date    delivered_delta  
0                        2017-10-18  -8 days +21:25:13  
1                        2018-08-13  -6 days +15:27:45  
2                        2018-09-04 -18 days +18:06:29  
3                        2017-12-15 -13 days +00:28:42  
4                        2018-02-26 -10 days +18:17:02  
...                             ...                ...  
99436                    2017-03-28 -11 days +15:08:01  
99437                    2018-03-02  -2 days +17:37:56  
99438                    2017-09-27  -6 days +11:24:17  
99439                    2018-02-15 -21 days +23:32:54  
99440                    2018-04-03 -18 days +13:08:30  

[99441 rows x 9 columns]

In [38]:
order_false_delivered = orders[orders['delivered_delta'] >= pd.Timedelta(days=1)] 

In [39]:
order_false_delivered

order_id                       customer_id  \
20     203096f03d82e0dffbc41ebc2e2bcfb7  d2b091571da224a1b36412c18bc3bbfe   
25     fbf9ac61453ac646ce8ad9783d7d0af6  3a874b4d4c4b6543206ff5d89287f0c3   
41     6ea2f835b4556291ffdc53fa0b3b95e8  c7340080e394356141681bd4c9b8fe31   
57     66e4624ae69e7dc89bd50222b59f581f  684fa6da5134b9e4dab731e00011712d   
58     a685d016c8a26f71a0bb67821070e398  911e4c37f5cafe1604fe6767034bf1ae   
...                                 ...                               ...   
99396  b3112ca67f3afd4e20cf2ee91fc4f804  6f83c71b6c044fb156d697d4130fe9b5   
99397  0fa1fab1d7c1211c824596ed5e111e3c  7f3bd6c94d2daf7b6462d1a894a775b4   
99423  38e9133ce29f6bbe35aed9c3863dce01  ad312389a098ceff46ce92c4595c06d0   
99425  d692ef54145c9cb3322ec2e5508aa3f4  82ddfcf9438b0cd1117b55ac33184df8   
99432  cfa78b997e329a5295b4ee6972c02979  a2f7428f0cafbc8e59f20e1444b67315   

      order_status order_purchase_timestamp   order_approved_at  \
20       delivered      2017-09-18 14:31:30 2017-09-19 04:04:09   
25       delivered      2018-02-20 23:46:53 2018-02-22 02:30:46   
41       delivered      2017-11-24 21:27:48 2017-11-25 00:21:09   
57       delivered      2018-03-09 14:50:15 2018-03-09 15:40:39   
58       delivered      2017-03-13 18:14:36 2017-03-13 18:14:36   
...            ...                      ...                 ...   
99396    delivered      2018-08-02 22:46:54 2018-08-02 23:04:06   
99397    delivered      2018-03-13 21:48:57 2018-03-13 22:40:28   
99423    delivered      2017-10-12 20:54:11 2017-10-14 03:28:24   
99425    delivered      2018-03-21 19:47:18 2018-03-21 20:05:26   
99432    delivered      2017-12-20 09:52:41 2017-12-20 10:09:52   

      order_delivered_carrier_date order_delivered_customer_date  \
20             2017-10-06 17:50:03           2017-10-09 22:23:46   
25             2018-02-26 22:25:22           2018-03-21 22:03:54   
41             2017-12-13 21:14:05           2017-12-28 18:59:23   
57             2018-03-15 00:31:19           2018-04-03 13:28:46   
58             2017-03-22 14:03:09           2017-04-06 13:37:16   
...                            ...                           ...   
99396          2018-08-15 17:42:00           2018-08-21 00:03:26   
99397          2018-03-14 19:27:23           2018-04-05 19:59:49   
99423          2017-10-17 17:04:42           2017-11-21 17:06:59   
99425          2018-03-22 21:11:58           2018-04-11 00:48:31   
99432          2017-12-20 20:25:25           2018-01-26 15:45:14   

      order_estimated_delivery_date  delivered_delta  
20                       2017-09-28 11 days 22:23:46  
25                       2018-03-12  9 days 22:03:54  
41                       2017-12-21  7 days 18:59:23  
57                       2018-04-02  1 days 13:28:46  
58                       2017-03-30  7 days 13:37:16  
...                             ...              ...  
99396                    2018-08-16  5 days 00:03:26  
99397                    2018-04-02  3 days 19:59:49  
99423                    2017-10-31 21 days 17:06:59  
99425                    2018-04-09  2 days 00:48:31  
99432                    2018-01-18  8 days 15:45:14  

[6535 rows x 9 columns]

In [40]:
order_false_delivered_ans = len(order_false_delivered)

In [41]:
order_false_delivered_ans

6535

In [42]:
print(f'Количество просроченных доставок заказов : {order_false_delivered_ans}')

Количество просроченных доставок заказов : 6535


In [43]:
#df['month'] = df['date_column'].dt.month преобразование даты к месяцу

In [44]:
dfs = [canceled_orders, unavailable_orders, order_false_delivered]

In [45]:
combined_df = pd.concat(dfs) #объеденили все заказы недоставленные по разным причинам или доставленные с просрочкой 

In [46]:
combined_df

order_id                       customer_id  \
397    1b9ecfe83cdc259250e1a8aca174f0ad  6d6b50b66d79f80827b6d96751528d30   
613    714fb133a6730ab81fa1d3c1b2007291  e3fe72696c4713d64d3c10afe71e75ed   
1058   3a129877493c8189c59c60eb71d97c29  0913cdce793684e52bbfac69d87e91fd   
1130   00b1cb0320190ca0daa2c88b35206009  3532ba38a3fd242259a514ac2b6ae6b6   
1801   ed3efbd3a87bea76c2812c66a0b32219  191984a8ba4cbb2145acb4fe35b69664   
...                                 ...                               ...   
99396  b3112ca67f3afd4e20cf2ee91fc4f804  6f83c71b6c044fb156d697d4130fe9b5   
99397  0fa1fab1d7c1211c824596ed5e111e3c  7f3bd6c94d2daf7b6462d1a894a775b4   
99423  38e9133ce29f6bbe35aed9c3863dce01  ad312389a098ceff46ce92c4595c06d0   
99425  d692ef54145c9cb3322ec2e5508aa3f4  82ddfcf9438b0cd1117b55ac33184df8   
99432  cfa78b997e329a5295b4ee6972c02979  a2f7428f0cafbc8e59f20e1444b67315   

      order_status order_purchase_timestamp    order_approved_at  \
397       canceled      2018-08-04 14:29:27  2018-08-07 04:10:26   
613       canceled      2018-01-26 21:34:08  2018-01-26 21:58:39   
1058      canceled      2018-01-25 13:34:24  2018-01-25 13:50:20   
1130      canceled      2018-08-28 15:26:39                  NaN   
1801      canceled      2018-09-20 13:54:16                  NaN   
...            ...                      ...                  ...   
99396    delivered      2018-08-02 22:46:54  2018-08-02 23:04:06   
99397    delivered      2018-03-13 21:48:57  2018-03-13 22:40:28   
99423    delivered      2017-10-12 20:54:11  2017-10-14 03:28:24   
99425    delivered      2018-03-21 19:47:18  2018-03-21 20:05:26   
99432    delivered      2017-12-20 09:52:41  2017-12-20 10:09:52   

      order_delivered_carrier_date order_delivered_customer_date  \
397                            NaN                           NaN   
613            2018-01-29 22:33:25                           NaN   
1058           2018-01-26 21:42:18                           NaN   
1130                           NaN                           NaN   
1801                           NaN                           NaN   
...                            ...                           ...   
99396          2018-08-15 17:42:00           2018-08-21 00:03:26   
99397          2018-03-14 19:27:23           2018-04-05 19:59:49   
99423          2017-10-17 17:04:42           2017-11-21 17:06:59   
99425          2018-03-22 21:11:58           2018-04-11 00:48:31   
99432          2017-12-20 20:25:25           2018-01-26 15:45:14   

      order_estimated_delivery_date  delivered_delta  
397             2018-08-14 00:00:00              NaT  
613             2018-02-22 00:00:00              NaT  
1058            2018-02-23 00:00:00              NaT  
1130            2018-09-12 00:00:00              NaT  
1801            2018-10-17 00:00:00              NaT  
...                             ...              ...  
99396           2018-08-16 00:00:00  5 days 00:03:26  
99397           2018-04-02 00:00:00  3 days 19:59:49  
99423           2017-10-31 00:00:00 21 days 17:06:59  
99425           2018-04-09 00:00:00  2 days 00:48:31  
99432           2018-01-18 00:00:00  8 days 15:45:14  

[7769 rows x 9 columns]

In [58]:
combined_df.dtypes

order_id                                  object
customer_id                               object
order_status                              object
order_purchase_timestamp                  object
order_approved_at                         object
order_delivered_carrier_date              object
order_delivered_customer_date             object
order_estimated_delivery_date             object
delivered_delta                  timedelta64[ns]
month                                     object
dtype: object

In [59]:
combined_df = combined_df.astype({'order_estimated_delivery_date':'datetime64',
                                'order_delivered_customer_date':'datetime64',
                                'order_delivered_carrier_date':'datetime64',
                                'order_approved_at': 'datetime64',
                                'order_purchase_timestamp': 'datetime64'})

In [60]:
combined_df.dtypes

order_id                                  object
customer_id                               object
order_status                              object
order_purchase_timestamp          datetime64[ns]
order_approved_at                 datetime64[ns]
order_delivered_carrier_date      datetime64[ns]
order_delivered_customer_date     datetime64[ns]
order_estimated_delivery_date     datetime64[ns]
delivered_delta                  timedelta64[ns]
month                                     object
dtype: object

In [61]:
combined_df['month'] = combined_df['order_purchase_timestamp'].dt.month

In [62]:
combined_df

order_id                       customer_id  \
397    1b9ecfe83cdc259250e1a8aca174f0ad  6d6b50b66d79f80827b6d96751528d30   
613    714fb133a6730ab81fa1d3c1b2007291  e3fe72696c4713d64d3c10afe71e75ed   
1058   3a129877493c8189c59c60eb71d97c29  0913cdce793684e52bbfac69d87e91fd   
1130   00b1cb0320190ca0daa2c88b35206009  3532ba38a3fd242259a514ac2b6ae6b6   
1801   ed3efbd3a87bea76c2812c66a0b32219  191984a8ba4cbb2145acb4fe35b69664   
...                                 ...                               ...   
99396  b3112ca67f3afd4e20cf2ee91fc4f804  6f83c71b6c044fb156d697d4130fe9b5   
99397  0fa1fab1d7c1211c824596ed5e111e3c  7f3bd6c94d2daf7b6462d1a894a775b4   
99423  38e9133ce29f6bbe35aed9c3863dce01  ad312389a098ceff46ce92c4595c06d0   
99425  d692ef54145c9cb3322ec2e5508aa3f4  82ddfcf9438b0cd1117b55ac33184df8   
99432  cfa78b997e329a5295b4ee6972c02979  a2f7428f0cafbc8e59f20e1444b67315   

      order_status order_purchase_timestamp   order_approved_at  \
397       canceled      2018-08-04 14:29:27 2018-08-07 04:10:26   
613       canceled      2018-01-26 21:34:08 2018-01-26 21:58:39   
1058      canceled      2018-01-25 13:34:24 2018-01-25 13:50:20   
1130      canceled      2018-08-28 15:26:39                 NaT   
1801      canceled      2018-09-20 13:54:16                 NaT   
...            ...                      ...                 ...   
99396    delivered      2018-08-02 22:46:54 2018-08-02 23:04:06   
99397    delivered      2018-03-13 21:48:57 2018-03-13 22:40:28   
99423    delivered      2017-10-12 20:54:11 2017-10-14 03:28:24   
99425    delivered      2018-03-21 19:47:18 2018-03-21 20:05:26   
99432    delivered      2017-12-20 09:52:41 2017-12-20 10:09:52   

      order_delivered_carrier_date order_delivered_customer_date  \
397                            NaT                           NaT   
613            2018-01-29 22:33:25                           NaT   
1058           2018-01-26 21:42:18                           NaT   
1130                           NaT                           NaT   
1801                           NaT                           NaT   
...                            ...                           ...   
99396          2018-08-15 17:42:00           2018-08-21 00:03:26   
99397          2018-03-14 19:27:23           2018-04-05 19:59:49   
99423          2017-10-17 17:04:42           2017-11-21 17:06:59   
99425          2018-03-22 21:11:58           2018-04-11 00:48:31   
99432          2017-12-20 20:25:25           2018-01-26 15:45:14   

      order_estimated_delivery_date  delivered_delta  month  
397                      2018-08-14              NaT      8  
613                      2018-02-22              NaT      1  
1058                     2018-02-23              NaT      1  
1130                     2018-09-12              NaT      8  
1801                     2018-10-17              NaT      9  
...                             ...              ...    ...  
99396                    2018-08-16  5 days 00:03:26      8  
99397                    2018-04-02  3 days 19:59:49      3  
99423                    2017-10-31 21 days 17:06:59     10  
99425                    2018-04-09  2 days 00:48:31      3  
99432                    2018-01-18  8 days 15:45:14     12  

[7769 rows x 10 columns]

In [63]:
grouped_df = combined_df.groupby(['month', 'order_status']).agg({'customer_id':'count'}).rename(columns={'customer_id': 'count_order'})

In [64]:
grouped_df

count_order
month order_status             
1     canceled               37
      delivered             425
      unavailable            58
2     canceled               91
      delivered             975
      unavailable            75
3     canceled               59
      delivered            1444
      unavailable            49
4     canceled               33
      delivered             457
      unavailable            14
5     canceled               53
      delivered             549
      unavailable            47
6     canceled               34
      delivered             166
      unavailable            28
7     canceled               69
      delivered             316
      unavailable            70
8     canceled              111
      delivered             515
      unavailable            39
9     canceled               37
      delivered             183
      unavailable            38
10    canceled               54
      delivered             189
      unavailable            65
11    canceled               37
      delivered             904
      unavailable            84
12    canceled               11
      delivered             411
      unavailable            42

In [71]:
grouped_df = grouped_df.drop_duplicates().reset_index()
pivot_table = grouped_df.pivot(index='month', columns='order_status', values='count_order')



In [72]:
pivot_table

order_status  canceled  delivered  unavailable
month                                         
1                 37.0      425.0         58.0
2                 91.0      975.0         75.0
3                 59.0     1444.0         49.0
4                 33.0      457.0         14.0
5                 53.0      549.0         47.0
6                 34.0      166.0         28.0
7                 69.0      316.0         70.0
8                111.0      515.0         39.0
9                  NaN      183.0         38.0
10                54.0      189.0         65.0
11                 NaN      904.0         84.0
12                11.0      411.0         42.0

In [73]:
mean_month_false = pivot_table.mean()

In [74]:
mean_month_false

order_status
canceled        55.20
delivered      544.50
unavailable     50.75
dtype: float64

# ВЫВОД. Всреднем по месяцам 52 отмененных заказов, 544 заказов полученных с задержкой,51 заказов недоступных

# 3 По каждому товару определить, в какой день недели товар чаще всего покупается.

In [87]:
orders['day_of_week'] = orders['order_purchase_timestamp'].dt.day_name()

In [88]:
orders

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp   order_approved_at  \
0        delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   
...            ...                      ...                 ...   
99436    delivered      2017-03-09 09:54:05 2017-03-09 09:54:05   
99437    delivered      2018-02-06 12:58:58 2018-02-06 13:10:37   
99438    delivered      2017-08-27 14:46:43 2017-08-27 15:04:16   
99439    delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
99440    delivered      2018-03-08 20:57:30 2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date    delivered_delta   day_week day_of_week  
0                        2017-10-18  -8 days +21:25:13     Monday      Monday  
1                        2018-08-13  -6 days +15:27:45    Tuesday     Tuesday  
2                        2018-09-04 -18 days +18:06:29  Wednesday   Wednesday  
3                        2017-12-15 -13 days +00:28:42   Saturday    Saturday  
4                        2018-02-26 -10 days +18:17:02    Tuesday     Tuesday  
...                             ...                ...        ...         ...  
99436                    2017-03-28 -11 days +15:08:01   Thursday    Thursday  
99437                    2018-03-02  -2 days +17:37:56    Tuesday     Tuesday  
99438                    2017-09-27  -6 days +11:24:17     Sunday      Sunday  
99439                    2018-02-15 -21 days +23:32:54     Monday      Monday  
99440                    2018-04-03 -18 days +13:08:30   Thursday    Thursday  

[99441 rows x 11 columns]

In [90]:
orders_with_products = pd.merge(orders, order_items, on='order_id', how='left')

In [91]:
orders_with_products #добавили product_id

order_id                       customer_id  \
0       e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1       53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2       47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3       949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4       ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                  ...                               ...   
113420  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
113421  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
113422  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
113423  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
113424  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

       order_status order_purchase_timestamp   order_approved_at  \
0         delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1         delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2         delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3         delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4         delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   
...             ...                      ...                 ...   
113420    delivered      2018-02-06 12:58:58 2018-02-06 13:10:37   
113421    delivered      2017-08-27 14:46:43 2017-08-27 15:04:16   
113422    delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
113423    delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
113424    delivered      2018-03-08 20:57:30 2018-03-09 11:20:28   

       order_delivered_carrier_date order_delivered_customer_date  \
0               2017-10-04 19:55:00           2017-10-10 21:25:13   
1               2018-07-26 14:31:00           2018-08-07 15:27:45   
2               2018-08-08 13:50:00           2018-08-17 18:06:29   
3               2017-11-22 13:39:59           2017-12-02 00:28:42   
4               2018-02-14 19:46:34           2018-02-16 18:17:02   
...                             ...                           ...   
113420          2018-02-07 23:22:42           2018-02-28 17:37:56   
113421          2017-08-28 20:52:26           2017-09-21 11:24:17   
113422          2018-01-12 15:35:03           2018-01-25 23:32:54   
113423          2018-01-12 15:35:03           2018-01-25 23:32:54   
113424          2018-03-09 22:11:59           2018-03-16 13:08:30   

       order_estimated_delivery_date    delivered_delta   day_week  \
0                         2017-10-18  -8 days +21:25:13     Monday   
1                         2018-08-13  -6 days +15:27:45    Tuesday   
2                         2018-09-04 -18 days +18:06:29  Wednesday   
3                         2017-12-15 -13 days +00:28:42   Saturday   
4                         2018-02-26 -10 days +18:17:02    Tuesday   
...                              ...                ...        ...   
113420                    2018-03-02  -2 days +17:37:56    Tuesday   
113421                    2017-09-27  -6 days +11:24:17     Sunday   
113422                    2018-02-15 -21 days +23:32:54     Monday   
113423                    2018-02-15 -21 days +23:32:54     Monday   
113424                    2018-04-03 -18 days +13:08:30   Thursday   

       day_of_week  order_item_id                        product_id  \
0           Monday            1.0  87285b34884572647811a353c7ac498a   
1          Tuesday            1.0  595fac2a385ac33a80bd5114aec74eb8   
2        Wednesday            1.0  aa4383b373c6aca5d8797843e5594415   
3         Saturday            1.0  d0b61bfb1de832b15ba9d266ca96e5b0   
4          Tuesday            1.0  65266b2da20d04dbe00c5c2d3bb7859e   
...            ...            ...                               ...   
113420     Tuesday            1.0  f1d4ce8c6dd66c47bbaa8c6781c2a923   
113421      Sunday            1.0  b8091

In [95]:
week_sale = orders_with_products.groupby(['product_id', 'day_of_week']).size().reset_index(name='count')

In [96]:
week_sale

product_id day_of_week  count
0      00066f42aeeb9f3007548bb9d3f33c38      Sunday      1
1      00088930e925c41fd95ebfe695fd2655     Tuesday      1
2      0009406fd7479715e4bef61dd91f2462    Thursday      1
3      000b8f95fcb9e0096488278317764d19      Friday      1
4      000b8f95fcb9e0096488278317764d19   Wednesday      1
...                                 ...         ...    ...
61225  fffdb2d0ec8d6a61f0a0a0db3f25b441      Friday      1
61226  fffdb2d0ec8d6a61f0a0a0db3f25b441      Sunday      1
61227  fffdb2d0ec8d6a61f0a0a0db3f25b441    Thursday      1
61228  fffdb2d0ec8d6a61f0a0a0db3f25b441     Tuesday      2
61229  fffe9eeff12fcbd74a2f2b007dde0c58   Wednesday      1

[61230 rows x 3 columns]

In [98]:
popular_day = week_sale.sort_values('count', ascending=False).drop_duplicates('product_id')



In [99]:
popular_day

product_id day_of_week  count
15902  422879e10f46682990de24d770e7f83d   Wednesday     93
36513  99a4788cb24856965c36a24e339b6058      Monday     92
41045  aca2eb7d00ea1a7b8ebd4e68314663af    Thursday     89
20127  53b36df67ebb7c41585e8d54d6772e08     Tuesday     76
13537  389d119b48cf3043d311335e499d9c6b    Thursday     67
...                                 ...         ...    ...
24030  6432d844f2be41bb6f2fd279e23415da      Friday      1
24009  6421ae053bb80dff3ceae1f217f54ed4      Monday      1
24010  642247ccf62585635cefa1bf09f59b94      Sunday      1
24022  642e125cbc752b4eeb8faf442319c39f    Thursday      1
61229  fffe9eeff12fcbd74a2f2b007dde0c58   Wednesday      1

[32951 rows x 3 columns]

# ВЫВОД в дф popular_day представлен каждый уникальный product_id и выведен день с максимальным кол-во заказов 

# 4 Сколько у каждого из пользователей в среднем покупок в неделю (по месяцам)? Не стоит забывать, что внутри месяца может быть не целое количество недель. Например, в ноябре 2021 года 4,28 недели. И внутри метрики это нужно учесть.

In [105]:
orders

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp   order_approved_at  \
0        delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   
...            ...                      ...                 ...   
99436    delivered      2017-03-09 09:54:05 2017-03-09 09:54:05   
99437    delivered      2018-02-06 12:58:58 2018-02-06 13:10:37   
99438    delivered      2017-08-27 14:46:43 2017-08-27 15:04:16   
99439    delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
99440    delivered      2018-03-08 20:57:30 2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date    delivered_delta   day_week day_of_week  
0                        2017-10-18  -8 days +21:25:13     Monday      Monday  
1                        2018-08-13  -6 days +15:27:45    Tuesday     Tuesday  
2                        2018-09-04 -18 days +18:06:29  Wednesday   Wednesday  
3                        2017-12-15 -13 days +00:28:42   Saturday    Saturday  
4                        2018-02-26 -10 days +18:17:02    Tuesday     Tuesday  
...                             ...                ...        ...         ...  
99436                    2017-03-28 -11 days +15:08:01   Thursday    Thursday  
99437                    2018-03-02  -2 days +17:37:56    Tuesday     Tuesday  
99438                    2017-09-27  -6 days +11:24:17     Sunday      Sunday  
99439                    2018-02-15 -21 days +23:32:54     Monday      Monday  
99440                    2018-04-03 -18 days +13:08:30   Thursday    Thursday  

[99441 rows x 11 columns]

In [108]:
# создаем столбец с годом и месяцем заказа
orders['order_approved_month'] = orders.order_approved_at.dt.strftime('%Y-%m')

In [109]:
orders

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp   order_approved_at  \
0        delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   
...            ...                      ...                 ...   
99436    delivered      2017-03-09 09:54:05 2017-03-09 09:54:05   
99437    delivered      2018-02-06 12:58:58 2018-02-06 13:10:37   
99438    delivered      2017-08-27 14:46:43 2017-08-27 15:04:16   
99439    delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
99440    delivered      2018-03-08 20:57:30 2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date    delivered_delta   day_week day_of_week  \
0                        2017-10-18  -8 days +21:25:13     Monday      Monday   
1                        2018-08-13  -6 days +15:27:45    Tuesday     Tuesday   
2                        2018-09-04 -18 days +18:06:29  Wednesday   Wednesday   
3                        2017-12-15 -13 days +00:28:42   Saturday    Saturday   
4                        2018-02-26 -10 days +18:17:02    Tuesday     Tuesday   
...                             ...                ...        ...         ...   
99436                    2017-03-28 -11 days +15:08:01   Thursday    Thursday   
99437                    2018-03-02  -2 days +17:37:56    Tuesday     Tuesday   
99438                    2017-09-27  -6 days +11:24:17     Sunday      Sunday   
99439                    2018-02-15 -21 days +23:32:54     Monday      Monday   
99440                    2018-04-03 -18 days +13:08:30   Thursday    Thursday   

      order_approved_month  
0                  2017-10  
1                  2018-07  
2                  2018-08  
3                  2017-11  
4                  2018-02  
...                    ...  
99436              2017-03  
99437              2018-02  
99438              2017-08  
99439              2018-01  
99440              2018-03  

[99441 rows x 12 columns]

In [110]:
# определяем количество дней в месяце и считаем кол-во недель в месяце
orders['count_day_weeks'] = (orders.order_approved_at.dt.days_in_month / 7).

In [111]:
orders

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp   order_approved_at  \
0        delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   
...            ...                      ...                 ...   
99436    delivered      2017-03-09 09:54:05 2017-03-09 09:54:05   
99437    delivered      2018-02-06 12:58:58 2018-02-06 13:10:37   
99438    delivered      2017-08-27 14:46:43 2017-08-27 15:04:16   
99439    delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
99440    delivered      2018-03-08 20:57:30 2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date    delivered_delta   day_week day_of_week  \
0                        2017-10-18  -8 days +21:25:13     Monday      Monday   
1                        2018-08-13  -6 days +15:27:45    Tuesday     Tuesday   
2                        2018-09-04 -18 days +18:06:29  Wednesday   Wednesday   
3                        2017-12-15 -13 days +00:28:42   Saturday    Saturday   
4                        2018-02-26 -10 days +18:17:02    Tuesday     Tuesday   
...                             ...                ...        ...         ...   
99436                    2017-03-28 -11 days +15:08:01   Thursday    Thursday   
99437                    2018-03-02  -2 days +17:37:56    Tuesday     Tuesday   
99438                    2017-09-27  -6 days +11:24:17     Sunday      Sunday   
99439                    2018-02-15 -21 days +23:32:54     Monday      Monday   
99440                    2018-04-03 -18 days +13:08:30   Thursday    Thursday   

      order_approved_month  count_day_weeks  
0                  2017-10         4.428571  
1                  2018-07         4.428571  
2                  2018-08         4.428571  
3                  2017-11         4.285714  
4                  2018-02         4.000000  
...                    ...              ...  
99436              2017-03         4.428571  
99437              2018-02         4.000000  
99438              2017-08         4.428571  
99439              2018-01         4.428571  
99440

In [123]:
orders_count = orders \
    .groupby(['customer_id', 'order_approved_month', 'count_day_weeks'], as_index=False) \
    .agg({'order_id': 'count'}) \
    .rename(columns={'order_id': 'orders_count'}) \
    .groupby('customer_id') \
    .agg({'orders_count': 'mean'}) \
    .rename(columns={'orders_count': 'orders_mean'})




In [124]:
orders_count

orders_mean
customer_id                                  
00012a2ce6f8dcda20d059ce98491703            1
000161a058600d5901f007fab4c27140            1
0001fd6190edaaf884bcaf3d49edf079            1
0002414f95344307404f0ace7a26f1d5            1
000379cdec625522490c315e70c7a9fb            1
...                                       ...
fffecc9f79fd8c764f843e9951b11341            1
fffeda5b6d849fbd39689bb92087f431            1
ffff42319e9b2d713724ae527742af25            1
ffffa3172527f765de70084a7e53aae8            1
ffffe8b65bbe3087b653a978c870db99            1

[99281 rows x 1 columns]

In [127]:
orders_count = orders_count.merge(orders[['customer_id', 'count_day_weeks']], on='customer_id', how='left')
orders_count['mean_orders'] = orders_count['orders_mean'] / orders_count['count_day_weeks']



In [128]:
orders_count

customer_id  orders_mean  count_day_weeks  \
0      00012a2ce6f8dcda20d059ce98491703            1         4.285714   
1      000161a058600d5901f007fab4c27140            1         4.428571   
2      0001fd6190edaaf884bcaf3d49edf079            1         4.000000   
3      0002414f95344307404f0ace7a26f1d5            1         4.428571   
4      000379cdec625522490c315e70c7a9fb            1         4.285714   
...                                 ...          ...              ...   
99276  fffecc9f79fd8c764f843e9951b11341            1         4.428571   
99277  fffeda5b6d849fbd39689bb92087f431            1         4.428571   
99278  ffff42319e9b2d713724ae527742af25            1         4.285714   
99279  ffffa3172527f765de70084a7e53aae8            1         4.285714   
99280  ffffe8b65bbe3087b653a978c870db99            1         4.428571   

       mean_orders  
0         0.233333  
1         0.225806  
2         0.250000  
3         0.225806  
4         0.233333  
...            ...  
99276     0.225806  
99277     0.225806  
99278     0.233333  
99279     0.233333  
99280     0.225806  

[99281 rows x 4 columns]

# ВЫВОД: в дф orders_count представленно среднее кол-во покупок в месяц в столбце mean_orders

# 5.1. Выполните когортный анализ пользователей.

In [135]:
df_cogort = pd.merge(orders, order_items, on='order_id', how='left')

In [136]:
df_cogort

order_id                       customer_id  \
0       e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1       53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2       47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3       949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4       ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                  ...                               ...   
113420  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
113421  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
113422  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
113423  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
113424  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

       order_status order_purchase_timestamp   order_approved_at  \
0         delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1         delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2         delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3         delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4         delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   
...             ...                      ...                 ...   
113420    delivered      2018-02-06 12:58:58 2018-02-06 13:10:37   
113421    delivered      2017-08-27 14:46:43 2017-08-27 15:04:16   
113422    delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
113423    delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
113424    delivered      2018-03-08 20:57:30 2018-03-09 11:20:28   

       order_delivered_carrier_date order_delivered_customer_date  \
0               2017-10-04 19:55:00           2017-10-10 21:25:13   
1               2018-07-26 14:31:00           2018-08-07 15:27:45   
2               2018-08-08 13:50:00           2018-08-17 18:06:29   
3               2017-11-22 13:39:59           2017-12-02 00:28:42   
4               2018-02-14 19:46:34           2018-02-16 18:17:02   
...                             ...                           ...   
113420          2018-02-07 23:22:42           2018-02-28 17:37:56   
113421          2017-08-28 20:52:26           2017-09-21 11:24:17   
113422          2018-01-12 15:35:03           2018-01-25 23:32:54   
113423          2018-01-12 15:35:03           2018-01-25 23:32:54   
113424          2018-03-09 22:11:59           2018-03-16 13:08:30   

       order_estimated_delivery_date    delivered_delta   day_week  \
0                         2017-10-18  -8 days +21:25:13     Monday   
1                         2018-08-13  -6 days +15:27:45    Tuesday   
2                         2018-09-04 -18 days +18:06:29  Wednesday   
3                         2017-12-15 -13 days +00:28:42   Saturday   
4                         2018-02-26 -10 days +18:17:02    Tuesday   
...                              ...                ...        ...   
113420                    2018-03-02  -2 days +17:37:56    Tuesday   
113421                    2017-09-27  -6 days +11:24:17     Sunday   
113422                    2018-02-15 -21 days +23:32:54     Monday   
113423                    2018-02-15 -21 days +23:32:54     Monday   
113424                    2018-04-03 -18 days +13:08:30   Thursday   

       day_of_week order_approved_month  count_day_weeks  order_item_id  \
0           Monday              2017-10         4.428571            1.0   
1          Tuesday              2018-07         4.428571            1.0   
2        Wednesday              2018-08         4.428571            1.0   
3         Saturday              2017-11         4.285714            1.0   
4          Tuesday              2018-02         4.000000            1.0   
...            ...                  ...              ...            ...   
113420     Tuesday              2018-02         4.000000            1.0   
113421  

In [140]:
first_order_month = df_cogort.groupby('customer_id')['order_approved_month'].min()
first_order_month.name = 'first_month' #Определение месяца первой покупки каждого пользователя

In [141]:
df_cogort = df_cogort.join(first_order_month, on='customer_id')

In [142]:
df_cogort

order_id                       customer_id  \
0       e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1       53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2       47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3       949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4       ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                  ...                               ...   
113420  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
113421  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
113422  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
113423  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
113424  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

       order_status order_purchase_timestamp   order_approved_at  \
0         delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1         delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2         delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3         delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4         delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   
...             ...                      ...                 ...   
113420    delivered      2018-02-06 12:58:58 2018-02-06 13:10:37   
113421    delivered      2017-08-27 14:46:43 2017-08-27 15:04:16   
113422    delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
113423    delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
113424    delivered      2018-03-08 20:57:30 2018-03-09 11:20:28   

       order_delivered_carrier_date order_delivered_customer_date  \
0               2017-10-04 19:55:00           2017-10-10 21:25:13   
1               2018-07-26 14:31:00           2018-08-07 15:27:45   
2               2018-08-08 13:50:00           2018-08-17 18:06:29   
3               2017-11-22 13:39:59           2017-12-02 00:28:42   
4               2018-02-14 19:46:34           2018-02-16 18:17:02   
...                             ...                           ...   
113420          2018-02-07 23:22:42           2018-02-28 17:37:56   
113421          2017-08-28 20:52:26           2017-09-21 11:24:17   
113422          2018-01-12 15:35:03           2018-01-25 23:32:54   
113423          2018-01-12 15:35:03           2018-01-25 23:32:54   
113424          2018-03-09 22:11:59           2018-03-16 13:08:30   

       order_estimated_delivery_date    delivered_delta   day_week  \
0                         2017-10-18  -8 days +21:25:13     Monday   
1                         2018-08-13  -6 days +15:27:45    Tuesday   
2                         2018-09-04 -18 days +18:06:29  Wednesday   
3                         2017-12-15 -13 days +00:28:42   Saturday   
4                         2018-02-26 -10 days +18:17:02    Tuesday   
...                              ...                ...        ...   
113420                    2018-03-02  -2 days +17:37:56    Tuesday   
113421                    2017-09-27  -6 days +11:24:17     Sunday   
113422                    2018-02-15 -21 days +23:32:54     Monday   
113423                    2018-02-15 -21 days +23:32:54     Monday   
113424                    2018-04-03 -18 days +13:08:30   Thursday   

       day_of_week order_approved_month  count_day_weeks  order_item_id  \
0           Monday              2017-10         4.428571            1.0   
1          Tuesday              2018-07         4.428571            1.0   
2        Wednesday              2018-08         4.428571            1.0   
3         Saturday              2017-11         4.285714            1.0   
4          Tuesday              2018-02         4.000000            1.0   
...            ...                  ...              ...            ...   
113420     Tuesday              2018-02         4.000000            1.0   
113421  

In [145]:
df_cogort = df_cogort.groupby(['first_month', 'order_purchase_timestamp']) \
              .agg({'customer_id': 'nunique'}).rename(columns={'customer_id': 'num_of_users'})

In [146]:
df_cogort

num_of_users
first_month order_purchase_timestamp              
2016-09     2016-09-15 12:16:38                  1
2016-10     2016-09-04 21:15:19                  1
            2016-09-05 00:15:34                  1
            2016-09-13 15:24:19                  1
            2016-10-02 22:07:52                  1
...                                            ...
2018-08     2018-08-29 14:18:23                  1
            2018-08-29 14:18:28                  1
            2018-08-29 14:52:00                  1
            2018-08-29 15:00:37                  1
2018-09     2018-09-03 09:06:57                  1

[98734 rows x 1 columns]

In [157]:
df_cogort['cohort_period'] = df_cogort.groupby('first_month').cumcount()
df_cogort = df_cogort.reset_index()

In [158]:
df_cogort

first_month order_purchase_timestamp  num_of_users  cohort_period
0         2016-09      2016-09-15 12:16:38             1              0
1         2016-10      2016-09-04 21:15:19             1              0
2         2016-10      2016-09-05 00:15:34             1              1
3         2016-10      2016-09-13 15:24:19             1              2
4         2016-10      2016-10-02 22:07:52             1              3
...           ...                      ...           ...            ...
98729     2018-08      2018-08-29 14:18:23             1           6582
98730     2018-08      2018-08-29 14:18:28             1           6583
98731     2018-08      2018-08-29 14:52:00             1           6584
98732     2018-08      2018-08-29 15:00:37             1           6585
98733     2018-09      2018-09-03 09:06:57             1              0

[98734 rows x 4 columns]

In [159]:
cohort_table = df_cogort.pivot(columns = 'cohort_period', index = 'first_month', values= 'num_of_users')
cohort_table = cohort_table.divide(cohort_table.iloc[:,0],axis=0)
cohort_table

cohort_period  0     1     2     3     4     5     6     7     8     9     \
first_month                                                                 
2016-09         1.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2016-10         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2016-12         1.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2017-01         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2017-02         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2017-03         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2017-04         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2017-05         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2017-06         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2017-07         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2017-08         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2017-09         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2017-10         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2017-11         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2017-12         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2018-01         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2018-02         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2018-03         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2018-04         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2018-05         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2018-06         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2018-07         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2018-08         1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2018-09         1.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

cohort_period  ...  7318  7319  7320  7321  7322  7323  7324  7325  7326  7327  
first_month    ...                                                              
2016-09        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2016-10        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2016-12        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2017-01        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2017-02        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2017-03        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2017-04        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2017-05        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2017-06        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2017-07        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2017-08        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2017-09        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2017-10        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2017-11        ...   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0  
2017-12        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2018-01        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2018-02        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2018-03        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2018-04        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2018-05        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2018-06        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2018-07        ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2018-08        ...   NaN   NaN   NaN   NaN   NaN   Na